client/serviceregistration/workload.go

```diff
@@ -23,6 +23,11 @@ type WorkloadServices struct {
 	// caused this registration.
 	JobID string
 
+	// NomadNamespace provides additional context for providers regarding which
+	// nomad namespace caused this registration.
+	// Note: this can be different to Namespace if the provider is not Nomad
+	NomadNamespace string
```

Comment:
This name is really awkward now that we have Nomad native service discovery, because there's a `Namespace` field that may _also_ be the Nomad namespace. We're passing a bunch of individual fields from the allocation runner, which don't have an obvious purpose until we get way down in the weeds here. These should probably get all wrapped up in some AllocInfo struct that we can instantiate in the alloc runner's group service hook.

(cc @jrasell and @shoenig on that though... would love your opinions on whether this smells like a bad suggestion to you)

Reply:
Yeah, I know. I didn't know how to call this. It felt weird setting that name there. I'll wait for suggestions on the changes here.

Reply:
If we go with adding a new field for the short-term, I'd vote for `JobNamespace` which aligns with `JobID` and would allow me to half-reason each namespace value has a slightly different meaning. This is wilful ignorance to the fact when using Nomad service discovery they are exactly the same though.

I would agree having an easier to reason about and understand structure to instantiate this information would be amazing. It is worth noting this would need to be available to both the AllocRunner and TaskRunner as both include service hooks for register, update, and deregister actions. Off the top of my head though, I can't quite envisage how best this would look.

Reply:
What if the existing `Namespace` field was renamed to `ProviderNamespace`, letting this field be name `Namespace`?

But yeah, I'm also on board with @tgross in just creating some kind of `AllocInfo` struct with the common alloc specific fields and embedding that into `WorkloadServices`

Reply:
I created a new struct called AllocInfo for this (in nomad/structs) and changed references all around (there a couple of changed files...). Also renamed the workload namespace to providerNamespace.